2-Draft-Accepted/RFCxxxx-Secrets-Management.md

```diff
@@ -0,0 +1,150 @@
+# from remote vaults
+$azDevOpsToken = Get-Secret -Name AzDevOps
+Invoke-RestMethod https://dev.azure.com/… -Credential $azDevOpsToken -Authentication Basic
+```
+
+Registering and using local secret:
+
+```powershell
+# For local vault, we can register custom secrets
+# In this example, we store a PSCredential object
+Register-Secret -Secret $cred -Name MyCreds
+New-PSSession -ComputerName myServer -Credential (Get-Secret -Name MyCreds)
+```
+
+## Specification
+
+### Secrets Vault
+
+There are two types of vaults: local and remote.
+A local vault, by default, is already created and named `Default`.
+`Default` on Windows is [CredMan](https://docs.microsoft.com/en-us/windows/desktop/SecAuthN/credentials-management).
+`Default` on macOS is [KeyChain](https://developer.apple.com/documentation/security/keychain_services).
+
+>[!NOTE]
+>KeyChain support is unlikely to be available in the first release of this feature.
+
+`Default` on Linux is [Hashicorp Vault](https://www.vaultproject.io/).
+
+>[!NOTE]
+>Linux has many options for local credential management, additional support may be
+>added in the future or via community provided extensions.
+
+### Credential Vault Extensions
+
+Vault extensions whether they are local or remote are provided as modules.
+
+Extensions are modules that provide a `Get-Secret` cmdlet that accepts a
+`-Credential` or `-AccessToken` parameter for authentication.
+Expectation is that each module will have their own cmdlet and not literally
+named `Get-Secret`.
+
+Extensions register themselves by calling `Register-SecretsVaultExtension`
+passing the name of the relevant cmdlet to `-Cmdlet`, the module name to `-Module`,
+and the module version to `-ModuleVersion`.
+This can be done as part of importing the module or explicitly by the user.
+
+`Register-SecretsVaultExtension` has a `-Default` switch to register the
+default local vault (if applicable, see above) if it has been unregistered.
+
+`Register-SecretsVaultExtension` takes a `-AccessToken` or `-Credential` parameter
+for authenticating with the remote vault.
+If neither are supplied, then the current user context is used and relies on
+the extension cmdlet to handle authentication.
+
+`Unregister-SecretsVaultExtension` should be used to remove any existing
+registration or when updating a registration to a new version.
+
+`Get-SecretsVaultExtension` will enumerate registered extensions returning
+the module name, module version, and cmdlet used.
+
+### Storing Secrets
+
+The `Register-Secret` cmdlet is used to store a secret.
+The `-Name` must be unique across all secrets regardless if local or remote.
+A `-Secret` parameter accepts an object that can be either a PSCredential
+or SecureString.
+
+### Retrieving Secrets
+
+The `Get-Secret` cmdlet is used to retrieve secrets.
+The `-Name` parameter is mandatory and retrieves the secret associated with
+that name.
+The returned object will either be a PSCredential or SecureString.
```

Comment:
Can we add a `-AsPlainText` switch? 

SecureStrings are notoriously hard to convert to plain strings, and the reality is that most non-PSCredential secrets need to be passed to APIs as plain strings. If you don't have a `-AsPlainText` people are going to be storing things as credentials just so they can use `.GetNetworkCredential().Password` to decode it.

Reply:
Since I'm adding support to just store strings, seems like securestring should only be used if they can be re-used as securestring so `-AsPlainText` is not necessary.

Reply:
I dunno about that.  I typically use `Read-Host -AsSecureString` to get my secret text into a variable so that it doesn't show up in my command history (or allow shoulder surfers to see it).   But what I really want is just the plain text for the invocation which requires plain text.  So I have to do the 
"create a pscredential / calll .GetNetworkCredential().Password" dance to decode it.   

Now if `Read-Host` had another parameter like `-HideInput` that only showed `*` as pasted the secret **and** gave me back a plain text string, then I'd be set.